ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java

```diff
@@ -290,15 +285,21 @@ public static BlockImportResult on_block(
     final BeaconState preState = store.getBlockState(block.getParent_root());
 
     // Return early if precondition checks fail;
-    final Optional<BlockImportResult> maybeFailure =
-        checkOnBlockPreconditions(block, preState, store);
-    if (maybeFailure.isPresent()) {
-      return maybeFailure.get();
+    final Optional<BlockImportResult> maybeFailureOnPreStorageConditions =
+        checkOnBlockPreStorageConditions(block, preState, store);
+    if (maybeFailureOnPreStorageConditions.isPresent()) {
+      return maybeFailureOnPreStorageConditions.get();
     }
 
     // Add new block to the store
     store.putBlock(block.hash_tree_root(), signed_block);
 
+    final Optional<BlockImportResult> maybeFailureOnPostStorageConditions =
+        checkOnBlockPostStorageConditions(block, store);
+    if (maybeFailureOnPostStorageConditions.isPresent()) {
+      return maybeFailureOnPostStorageConditions.get();
+    }
```

Comment:
The aim of these checks is to bail out before any processing is done if we don't have the block ancestor or other prerequisites.  We don't commit the transaction so any changes made to the store are discarded if we fail either of these checks.  So we should just combine them into one again.

Reply:
get_ancestor throws a NullPointerException if we combine these condition checks.

Reply:
Spec explicitly adds the block to the store before doing the 'postStorageConditionCheck' as I call here.

Reply:
Unless we add the block that is being processed here to the store, get_ancestor does not have access to this block and thus throws the NPE. In order to combine these checks and still not get the NPE, we'd need to change get_ancestor, which doesn't seem to be worth the trade-off especially since most of the fork choice is going to be modified during the protoArray change.

Reply:
Yeah, the spec has always added before doing all the checks and we deliberately moved away from that. :)

We already know that we have the block's parent on the chain, so we can just call `get_ancestor(store, block.getParent_root(), finalizedSlot);`.  Which is what we used to have.  This check hasn't actually changed, the spec just checks the slot first then the get_ancestor call whereas previously it did it the other way around.

Reply:
Alright reverting. That makes sense.